src/lib/backward-search/backward-search.ts

```diff
@@ -1,151 +1,294 @@
-import { LearningUnit, Skill } from "../..";
+import { isSkill, LearningUnit, PotentialNode, Skill, Unit } from "../..";
+import { GlobalKnowledge } from "../fastDownward/global-knowledge";
+import { Graph as GraphLib } from "@dagrejs/graphlib";
 
 /**
- * Filter the learning units to reduce the size of the potential candidate for the algorithm.
- * Tracing backward the potential learning units from the goal using required skills.
- * Taking in consideration the parent/children relation.
+ * Analysis skills of a goal to find the missing skills.
+ * Create a potential graph using createGoalsGraph function.
+ * The algorithm traces back each path in the a potential graph from the goal backward, recursively.
+ * Get all successors (children) of a node, then get all successors (children) of each them and so on, recursively.
+ *
+ * Skill is considered a missing skill if:
+ *      A) A path ends with with this skill
+ *      AND
+ *      B) The parent of the skill does not have a learning unit as child,
+ *         Meaning that the parent can not be learned through learning units (parent/child).
+ *
+ * Example:
+ *                => Sk:02 => lu:02 => Sk:04 => lu:04
+ *                |
+ * Sk:01 => Lu:01 |
+ *                |                 => Sk:05 => lu:05
+ *                => Sk:03 => lu:03 |
+ *                                  => Sk:06
+ *
+ * Path A) Sk:1 => Lu:01 => Sk:02 => lu:02 => Sk:04 => lu:04 (There are no Missing Skills)
+ * Path B) Sk:1 => Lu:01 => Sk:03 => lu:03 => Sk:05 => lu:05 (There are no Missing Skills)
+ * Path C) Sk:1 => Lu:01 => Sk:03 => lu:03 => Sk:06 ("Sk:06" is considered a Missing Skill)
+ *
+ * @param goals The skills that should be learned.
+ * @param allUnits The set of all LearningUnits.
+ * @param allSkills The set of all skills (independent of what was already learned and what should be learned).
+ * @param knowledge The knowledge of the user (skills already learned).
+ * @returns A list of PotentialNodes.
+ */
+export function skillAnalysis<LU extends LearningUnit>(
+    goals: Skill[],
+    allUnits: ReadonlyArray<Unit<LU>>,
+    allSkills: ReadonlyArray<Skill>,
+    knowledge: Skill[]
+): PotentialNode<LU>[] {
+    // Create a potential graph that holds the potential learning units and skills
+    const graph = createGoalsGraph(goals, allUnits, allSkills, knowledge);
```

Comment:
Is graph-based approach realy faster? I thought about an algorithm with ~3 lists:
* `goal-list`: One list storing the goals to learn
* `processed-list`: A list storing the learned goals to avoid re-processing, can also store the current knowledge
* `units`: List of used LearningUnits
Backward search:
1. Add all units that provide at least one element from the goal list
  1. Move all provided goals of the units from `goal-list` to `processed-list`
  2. Add all requirements that are not on `processed-list` to goal-list
  3. Add LearningUnit to `units`
2. Repeat until `goal-list` is empty.

This approach does not require computation of edges and nodes as required by a graph algorithm.

Reply:
I used a graph-based approach because I tried to use one backward search for both cycles and missing skills features.
The ~3 lists algorithm works fine with missing features, but I did not help to find cycles.

Reply:
Please check if the list approach is much faster than the graph based approch:

- If yes (at least a magnitude): Use grapbased for analysis (find cycles), list based for filtering when running the planner
- If not keep it as is

Reply:
I checked the performance of the graph and list backward search (previous implementation of the backward filter https://github.com/e-Learning-by-SSE/nm-skill-lib/blob/main/src/lib/backward-search/backward-search.ts)

I used the stress test with 5,000 skills and 5,000 learning units. The difference between graph and list backward search was less than 10 milliseconds.